### PR TITLE
#7386: Add better validation for sharded buffers

### PR DESCRIPTION
**Pull request by TT-BrianLiu**
- Add more stringent checks for row/col for BLOCK sharding
- Add fix to softmax and groupnorm for HEIGHT sharding
- Update silu to respect new BLOCK sharding checks
- Fix incorrect shard grids for ROW_MAJOR in test_sharded_tensor
- Fix incorrect strategy and shard spec in reshard tests
- Add proper program cache test for reshard op
- Remove manual override hash function from reshard op
- Fix nlp_concat_heads grid and fix COL_MAJOR for sharded_to_from_l1 test

### CODE REVIEW

File: tests/tt_eager/python_api_testing/unit_testing/misc/test_sharded.py [hunk 1]
```diff
@@ -84,14 +85,15 @@ def test_sharded_tile(
...
 
 
+# TODO (7735): Switch to new interleaved_to_sharded with sharded_mem_config input and re-enable BLOCK sharded tests
 @pytest.mark.parametrize(
     "input_shape, shard_scheme, shard_size, num_cores",
     [
         ([1, 1, 100352, 64], ttl.tensor.TensorMemoryLayout.HEIGHT_SHARDED, (1024, 64), 98),
         ([1, 1, 128, 50176], ttl.tensor.TensorMemoryLayout.WIDTH_SHARDED, (128, 512), 98),
-        ([1, 1, 100352, 64], ttl.tensor.TensorMemoryLayout.BLOCK_SHARDED, (2048, 32), 98),
+        # ([1, 1, 100352, 64], ttl.tensor.TensorMemoryLayout.BLOCK_SHARDED, (2048, 32), 98),
```
Comment (xanderchin) [inline tests/tt_eager/python_api_testing/unit_testing/misc/test_sharded.py:94]:
```
pytest.param(([1, 1, 100352, 64], ttl.tensor.TensorMemoryLayout.BLOCK_SHARDED, (2048, 32), 98, marks=pytest.mark.xfail(reason="7735: Switch to new interleaved_to_sharded with sharded_mem_config input and re-enable BLOCK sharded tests")),
```

**Review by xanderchin — APPROVED** [review]
Minor comments

File: tests/tt_eager/python_api_testing/sweep_tests/pytests/tt_dnn/test_softmax_sharded.py [hunk 2]
```diff
@@ -100,7 +100,7 @@ def test_softmax(device, in_dtype, causal_mask, grid_size, seq_len, scale_mask):
         in1_t,
         grid_size,
         [M // grid_size[1], K // grid_size[0]],
-        ttl.tensor.TensorMemoryLayout.BLOCK_SHARDED,
+        ttl.tensor.TensorMemoryLayout.HEIGHT_SHARDED,
```
[the remark this answers is not in the document]
Comment (TT-BrianLiu) [reply tests/tt_eager/python_api_testing/sweep_tests/pytests/tt_dnn/test_softmax_sharded.py:103]: likely not. I am changing BLOCK sharding to not be allowed when it's just HEIGHT sharding.

**Comment by TT-BrianLiu** [comment]
Passing pipeline: https://github.com/tenstorrent/tt-metal/actions/runs/8793775520

**Comment by TT-BrianLiu** [comment]
Passing pipelines after rebase: https://github.com/tenstorrent/tt-metal/actions/runs/8794104046

**Pull request merged**